// nestjs-schedule/dist/schedule-metadata.accessor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SchedulerMetadataAccessor = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const core_1 = require("@nestjs/core");
const schedule_constants_1 = require("./schedule.constants");
let SchedulerMetadataAccessor = class SchedulerMetadataAccessor {
    constructor(reflector) {
        this.reflector = reflector;
    }
    getSchedulerType(target) {
        return this.reflector.get(schedule_constants_1.SCHEDULER_TYPE, target);
    }
    getSchedulerName(target) {
        return this.reflector.get(schedule_constants_1.SCHEDULER_NAME, target);
    }
    getLocker(target) {
        return this.reflector.get(schedule_constants_1.SCHEDULE_LOCKER, target);
    }
    getJobOptions(target) {
        return this.reflector.get(schedule_constants_1.SCHEDULER_OPTIONS, target);
    }
    getTimeoutMetadata(target) {
        return this.reflector.get(schedule_constants_1.SCHEDULE_TIMEOUT_OPTIONS, target);
    }
    getIntervalMetadata(target) {
        return this.reflector.get(schedule_constants_1.SCHEDULE_INTERVAL_OPTIONS, target);
    }
    getCronMetadata(target) {
        return this.reflector.get(schedule_constants_1.SCHEDULE_CRON_OPTIONS, target);
    }
};
SchedulerMetadataAccessor = tslib_1.__decorate([
    (0, common_1.Injectable)(),
    tslib_1.__metadata("design:paramtypes", [core_1.Reflector])
], SchedulerMetadataAccessor);
exports.SchedulerMetadataAccessor = SchedulerMetadataAccessor;
